File: app/nodes/standard/knowledge/knowledge.js
function NodeDatastore() {
    this.addInput("Query", "string");
    this.addOutput("Context", "string");
    this.properties = { datastore: "", top_k: 3 };
    
    this.dsWidget = this.addWidget("combo", "Store", this.properties.datastore, (v) => { 
        this.properties.datastore = v; 
        if(window.pushHistoryState) window.pushHistoryState();
    }, { values: window.datastores_list || [] });

    this.kWidget = this.addWidget("number", "Top K", this.properties.top_k, (v) => { 
        this.properties.top_k = v; 
        if(window.pushHistoryState) window.pushHistoryState();
    }, { min: 1, max: 20 });

    this.addWidget("button", "ℹ️ Documentation", null, () => { showNodeHelp(this.type); });

    this.title = "📚 RAG DATASTORE";
    this.color = "#0d9488";
    this.serialize_widgets = true;
    this.size = this.computeSize();
}
NodeDatastore.title = "📚 RAG DATASTORE";
NodeDatastore.prototype.onConfigure = function(config) {
    if (this.dsWidget) this.dsWidget.value = this.properties.datastore;
    if (this.kWidget) this.kWidget.value = this.properties.top_k;
    if (!this.title || this.title === "NodeDatastore") this.title = "📚 RAG DATASTORE";
};
LiteGraph.registerNodeType("hub/datastore", NodeDatastore);

function NodeWebSearch() {
    this.addInput("Query", "string");
    this.addOutput("Results", "string");
    this.properties = { service: "wikipedia", max_results: 5 };
    
    this.sWidget = this.addWidget("combo", "Service", this.properties.service, (v) => { 
        this.properties.service = v; 
        if(window.pushHistoryState) window.pushHistoryState();
    }, { values: ["wikipedia", "arxiv", "google"] });
    
    this.mWidget = this.addWidget("number", "Max Results", this.properties.max_results, (v) => { 
        this.properties.max_results = v; 
        if(window.pushHistoryState) window.pushHistoryState();
    }, { min: 1, max: 10 });

    this.addWidget("button", "ℹ️ Documentation", null, () => { showNodeHelp(this.type); });

    this.title = "🌐 WEB SEARCH";
    this.color = "#3b82f6";
    this.serialize_widgets = true;
    this.size = this.computeSize();
}
NodeWebSearch.title = "🌐 WEB SEARCH";
NodeWebSearch.prototype.onConfigure = function() {
    if (this.sWidget) this.sWidget.value = this.properties.service;
    if (this.mWidget) this.mWidget.value = this.properties.max_results;
    if (!this.title || this.title === "NodeWebSearch") this.title = "🌐 WEB SEARCH";
};
LiteGraph.registerNodeType("hub/web_search", NodeWebSearch);

File: app/nodes/standard/knowledge/knowledge.py
from typing import Dict, Any
from sqlalchemy import select
from app.nodes.base import BaseNode
from app.database.models import DataStore
from app.database.session import AsyncSessionLocal
from app.core import knowledge_importer as kit
from fastapi.concurrency import run_in_threadpool

class RAGDatastoreNode(BaseNode):
    node_type = "hub/datastore"
    node_title = "RAG Datastore"
    node_category = "Knowledge & RAG"
    node_icon = "📚"

    async def execute(self, engine, node: Dict[str, Any], output_slot_idx: int) -> Any:
        import pipmaster as pm
        pm.ensure_packages(["safe-store"])
        from safe_store import SafeStore
        
        query = await engine._resolve_input(node, 0)
        store_name = node["properties"].get("datastore")
        top_k = int(node["properties"].get("top_k", 3))
        
        if not query or not store_name: return ""
        
        cb = getattr(engine.request.state, "stream_callback", None)
        if cb:
            await cb(f'* DATASTORE: Searching "{store_name}" for "{str(query)[:40]}..."\n')

        async with AsyncSessionLocal() as db:
            res = await db.execute(select(DataStore).filter(DataStore.name == store_name))
            ds = res.scalars().first()
            
        if not ds: return f"[Error: Datastore '{store_name}' not found]"

        V_MAP = {"sentense_transformer": "st", "tf_idf": "tf_idf", "ollama": "ollama", "openai": "openai", "cohere": "cohere", "lollms": "lollms"}
        v_key = V_MAP.get(ds.vectorizer_name, "tf_idf")

        def _query():
            s = SafeStore(db_path=ds.db_path, vectorizer_name=v_key, vectorizer_config=ds.vectorizer_config or {})
            with s: return s.query(str(query), top_k=top_k)

        raw_results = await run_in_threadpool(_query)
        context_parts = [f"[[SOURCE {i+1}: {r.get('document_title')}]]\n{r.get('chunk_text')}" for i, r in enumerate(raw_results)]
            
        return "\n\n---\n\n".join(context_parts)

class WebSearchNode(BaseNode):
    node_type = "hub/web_search"
    node_title = "Web Search"
    node_category = "Knowledge & RAG"
    node_icon = "🌐"

    async def execute(self, engine, node: Dict[str, Any], output_slot_idx: int) -> Any:
        query = await engine._resolve_input(node, 0)
        if not query: return ""
        service = node["properties"].get("service", "wikipedia")
        count = int(node["properties"].get("max_results", 5))

        cb = getattr(engine.request.state, "stream_callback", None)
        if cb:
            await cb(f'* WEB SEARCH: Querying {service.capitalize()} for "{str(query)[:40]}..."\n')

        if service == "wikipedia": res = await run_in_threadpool(kit.search_wikipedia_sync, str(query))
        elif service == "arxiv": res = await run_in_threadpool(kit.search_arxiv_sync, str(query), max_results=count)
        elif service == "google": res = await run_in_threadpool(kit.search_google_sync, str(query))
        else: res =[]

        parts = [f"[[WEB: {r['title']}]]\n{r['content']}" for r in res[:count]]
        return "\n\n---\n\n".join(parts)